feat: defined type flow rules and added test cases

aagarwal1012 opened the pull request:
@wmdietl, finally the basic test is working now. Please review this PR.
> I have overridden the `visitMemberSelect()` method to add the `@XXXContainer` annotation to `R.XXX` member select and the `@YYYRes` annotation to `R.XXX.YYY` member select.

aagarwal1012 committed:
fix: add XXXContainer qualifers to ResourceBottom

aagarwal1012 committed:
fix: convert wildcard imports into single imports

aagarwal1012 committed:
feat: added AnnotatedTypeFactory for android resources

aagarwal1012 committed:
rfac: convert if statements to switch in AndroidResourceAnnotatedTypeFactory

aagarwal1012 committed:
conf: added CF local dependency

aagarwal1012 committed:
feat: added basic test for android resource annotations

aagarwal1012 committed:
feat: added AndroidResourceVisitor

aagarwal1012 committed:
bfix: fixed NPE while testing

aagarwal1012 committed:
code: changed tests

aagarwal1012 committed:
rfac: merging master

aagarwal1012 committed:
feat: fixed visitMemberSelect to add annotations to Container and Res

aagarwal1012 committed:
git: fixed conflicts

aagarwal1012 committed:
feat: added Visitor that gives warning on binary and compound assignment

aagarwal1012 committed:
chor: removed @Resource annotation

aagarwal1012 committed:
Merge branch 'develop' into visitor

aagarwal1012 committed:
ptch: added @SupressWarnings to test

aagarwal1012 committed:
chor: commented print statements

aagarwal1012 committed:
Merge branch 'visitor' into testing

aagarwal1012 committed:
chor: added resources in R.java

wmdietl commented on src/main/java/org/checkerframework/checker/androidresource/qual/ResourceTop.java:
```diff
@@ -3,10 +3,7 @@
-import java.lang.annotation.Retention;
-import java.lang.annotation.RetentionPolicy;
-import java.lang.annotation.Target;
+import java.lang.annotation.*;
```
Can you undo all these wildcard import changes? In general explicit imports are less error prone.
It also makes it hard to see what actually changed in this PR.

aagarwal1012 replied:
I will do.

wmdietl commented on tests/androidresource/AndroidResourceTest.java:
```diff
@@ -0,0 +1,12 @@
+        @XmlRes int xml_res = R.xml.file_paths;
+//        xml_res = R.string.file_paths;    // Flagged by CF
+    }
+}
```
Add the newline at the end of files, for consistency and to get rid of the ugly red icon.

aagarwal1012 replied:
This will be resolved in #4

aagarwal1012 replied:
Resolved.

wmdietl commented on tests/androidresource/AndroidResourceTest.java:
```diff
@@ -0,0 +1,12 @@
+
+    void Test() {
+        @XmlRes int xml_res = R.xml.file_paths;
+//        xml_res = R.string.file_paths;    // Flagged by CF
```
Instead of a comment you must actually make this a test case.
See https://github.com/typetools/checker-framework/blob/master/checker/tests/README for detailed instructions.

aagarwal1012 replied:
This will be resolved in #4

wmdietl commented on .travis.yml:
```diff
@@ -1,3 +0,0 @@
-language: java
```
Why are you disabling travis completely? It would be better to actually enable running the tests on travis, so that for future PRs we will actually see whether they break existing tests.

aagarwal1012 replied:
Actually, I need to set up Travis for the repo, currently, it is just a dummy script which is failing. So I removed it.

aagarwal1012 replied:
@wmdietl, I have added Travis-ci in #4

wmdietl commented on build.gradle:
```diff
@@ -13,7 +13,10 @@ sourceCompatibility = 1.8
     // This dependency is found on compile classpath of this component and consumers.
-    implementation 'org.checkerframework:checker:2.8.2git'
+//    implementation 'org.checkerframework:checker:2.8.2'
+    compile files('E:/gsoc-2019/The-Checker-Framework/checker-framework/checker/dist/checker-qual.jar')
```
You should not commit local directories to the general configuration.
You can add a flag to run with a local CF Build, like you did for the template checker.

aagarwal1012 replied:
@wmdietl, there is some issue in windows so I have to give an explicit path to the jar files.

wmdietl commented on src/main/java/org/checkerframework/checker/androidresource/AndroidResourceAnnotatedTypeFactory.java:
```diff
@@ -0,0 +1,276 @@
+         * @param annotatedTypeMirror
+         * @param expression
+         */
+        private void assignResAnnotations(AnnotatedTypeMirror annotatedTypeMirror, String expression) {
```
This code is not using the `@XXXContainer` annotations to make a decision.
Again, there was a reason why we had the container annotations and you should write tests for those and then actually implement logic that uses the container annotations.

aagarwal1012 replied:
Sorry! I will update it soon. Thanks.

aagarwal1012 committed:
rfac: undo wildcard imports

aagarwal1012 committed:
rfac: made assignment of @XXXRes depend upon @XXXContainer

aagarwal1012 committed:
docs: added docs in AndroidResourceTypeFactory

aagarwal1012 committed:
Merge branch 'develop' into visitor

aagarwal1012 committed:
rfac: replace | to || in AndroidResourceVisitor

aagarwal1012 committed:
feat: added messages.properties

aagarwal1012 committed:
rfac: updated check res annotation method

aagarwal1012 committed:
Merge branch 'visitor' into testing

aagarwal1012 committed:
conf: changed build.gradle and added .travis.yml

aagarwal1012 committed:
bfix: added permission for travis

aagarwal1012 committed:
conf: added circle-ci

aagarwal1012 committed:
conf: removed circle-ci

aagarwal1012 committed:
bfix: fixed travis

aagarwal1012 committed:
bfix: fixed tests

aagarwal1012 committed:
ptch: added init test

aagarwal1012 committed:
ptch: added AnyRes annotation tests

aagarwal1012 committed:
chor: added a line after the end of each file

aagarwal1012 committed:
ptch: added other res tests

aagarwal1012 committed:
chor: added exception for EQUAL_TO and NOT_EQUAL_TO binary operation

aagarwal1012 committed:
Merge branch 'visitor' into testing

aagarwal1012 committed:
ptch: added tests for equal_to and not_equal_to binary operation

wmdietl commented:
@aagarwal1012 It's very hard to review this PR with so many comments pointing to #4. How about merging #4 into this one, merging all the simple import changes without review, and making a new PR that only contains the interesting changes that actually need a review?
It's also not clear why you split off #3 from this one. Can you prepare a more logic set of PRs?

aagarwal1012 commented:
@wmdietl, should I merge #3 into this PR?
Also, #4 only includes issues related to the test cases and Travis setup.

wmdietl commented:
Yes, merge #3 and #4 into this PR. The features and test cases should be in a single PR. Otherwise, how can we tell whether the feature is implemented correctly?
Split off the trivial changes to imports into a separate PR.

aagarwal1012 committed:
Merge pull request #4 from aagarwal1012/testing

Adding test cases

aagarwal1012 committed:
Merge pull request #3 from aagarwal1012/visitor

feat: added AndroidResourceVisitor

aagarwal1012 commented:
@wmdietl, please have a look at this PR.

wmdietl commented:
This still contains trivial changes to imports. Please make those in a separate PR and merge them.

aagarwal1012 committed:
Merge branch 'master' into develop

aagarwal1012 commented:
@wmdietl, I have updated the master branch. Please review this PR once again.

aagarwal1012 committed:
rfac: replaced @ImplicitFor to @DefaultFor

Done to match master branch of checker framework.

wmdietl commented on src/main/java/org/checkerframework/checker/androidresource/AndroidResourceVisitor.java:
```diff
@@ -0,0 +1,103 @@
+
+        if (annotationMirror == null)
+            return false;
+        else if (annotationMirror.equals(RESOURCE_TOP) || annotationMirror.equals(RESOURCE_BOTTOM))
```
Don't use `equals` to compare `AnnotationMirrors`. Use `AnnotationUtils.areSame`.
https://github.com/typetools/checker-framework/blob/master/javacutil/src/main/java/org/checkerframework/javacutil/AnnotationUtils.java#L91

aagarwal1012 replied:
Fixed.

aagarwal1012 committed:
rfac: changed annotation.equals to AnnotationUtils.areSame

wmdietl commented on src/main/java/org/checkerframework/checker/androidresource/AndroidResourceAnnotatedTypeFactory.java:
```diff
@@ -0,0 +1,402 @@
+            // checks if {expressionContainerAnnotation} is not null.
+            if (expressionContainerAnnotation != null) {
+
+                if (expressionContainerAnnotation.equals(ANIMATOR_CONTAINER)) {
```
You should replace all occurrences of `equals` on `AnnotationMirrors`.

aagarwal1012 replied:
Fixed.

wmdietl commented on build.gradle:
```diff
@@ -13,7 +13,10 @@ sourceCompatibility = 1.8
     // This dependency is found on compile classpath of this component and consumers.
-    implementation 'org.checkerframework:checker:2.8.2git'
+    implementation 'org.checkerframework:checker:2.8.2'
+//    compile files('E:/gsoc-2019/The-Checker-Framework/checker-framework/checker/dist/checker-qual.jar')
```
Instead of having these local comments, can you add the proper configuration to run using a locally-built version of the Checker Framework?

aagarwal1012 replied:
@wmdietl,
I am using CF local repo for the checker, I need to give absolute path due to some problem with Windows. Also, I have added maven CF dependency for the Travis tests. Currently, I need to figure out how to build android-resource-checker from CF repo in Travis.

wmdietl replied:
How about you keep this file clean and add the local configuration in a separate PR.

wmdietl commented on .travis.yml:
```diff
@@ -1,3 +1,5 @@
+dist: Trusty
+jdk: openjdk8
+script:
+  - ./gradlew test --info
```
In general, red icons are usually a bad sign. Make sure that your files have a newline before the end-of-file to avoid these.

aagarwal1012 replied:
Fixed.

wmdietl commented on src/main/java/org/checkerframework/checker/androidresource/AndroidResourceAnnotatedTypeFactory.java:
```diff
@@ -0,0 +1,402 @@
+
+import javax.lang.model.element.AnnotationMirror;
+
+public class AndroidResourceAnnotatedTypeFactory extends BaseAnnotatedTypeFactory {
```
Add some top-level documentation to all your classes and methods.

aagarwal1012 replied:
Done. Please refer #7 for docs related to annotation classes.

wmdietl commented on src/main/java/org/checkerframework/checker/androidresource/AndroidResourceAnnotatedTypeFactory.java:
```diff
@@ -0,0 +1,402 @@
+            AnnotationMirror expressionContainerAnnotation = expressionAnnotatedTypeMirror.getAnnotationInHierarchy(RESOURCE_TOP);
+
+            // checks if {expressionContainerAnnotation} is not null.
+            if (expressionContainerAnnotation != null) {
```
You can reduce nesting of your code if you change this to:

```
if (xxx == null) {
  return;
}
```
That makes the code a bit easier to read.

aagarwal1012 replied:
Fixed.

wmdietl commented on src/main/java/org/checkerframework/checker/androidresource/AndroidResourceAnnotatedTypeFactory.java:
```diff
@@ -0,0 +1,402 @@
+        @Override
+        public Void visitMemberSelect(MemberSelectTree node, AnnotatedTypeMirror annotatedTypeMirror) {
+
+            String expression = node.getExpression().toString();
```
Why not move these two variables into `assignContainerAnnotations` and make it a more local decision how it is implemented?

aagarwal1012 replied:
Fixed.

wmdietl commented on src/main/java/org/checkerframework/checker/androidresource/AndroidResourceAnnotatedTypeFactory.java:
```diff
@@ -0,0 +1,402 @@
+         * Assigning @XXXContainer annotations to member select.
+         */
+        private void assignContainerAnnotations(AnnotatedTypeMirror annotatedTypeMirror, String expression, String identifier) {
+            if (expression.equals("R")) {
```
It would again make the code a bit easier to read if you just return if the receiver is not `R`.

aagarwal1012 replied:
Fixed.

aagarwal1012 replied:
Fixed.

wmdietl commented on src/main/java/org/checkerframework/checker/androidresource/AndroidResourceAnnotatedTypeFactory.java:
```diff
@@ -0,0 +1,402 @@
+        public Void visitMemberSelect(MemberSelectTree node, AnnotatedTypeMirror annotatedTypeMirror) {
+
+            String expression = node.getExpression().toString();
+            String identifier = node.getIdentifier().toString();
```
`node.getIdentifier()` returns a `Name` object, which implements `CharSequence`. Isn't that good enough for the switch you perform on this value? I suspect you can safe yourself the `toString` call.

aagarwal1012 replied:
I think that it will make the code more difficult to read.

wmdietl commented on src/main/java/org/checkerframework/checker/androidresource/AndroidResourceVisitor.java:
```diff
@@ -0,0 +1,104 @@
+
+        AnnotationMirror annotationMirror = annotatedTypeMirror.getAnnotationInHierarchy(RESOURCE_TOP);
+
+        if (annotationMirror == null)
```
You can simplify these conditions to a single `return` statement. That would make the logic easier to understand.

aagarwal1012 replied:
I have added a variable `hasAnnotation` that will make the logic easier to understand.

wmdietl replied:
Isn't `return annotationMirror != null && !AnnotationUtils.areSame(annotationMirror, RESOURCE_TOP) && !AnnotationUtils.areSame(annotationMirror, RESOURCE_BOTTOM);` much easier?

wmdietl commented on src/main/java/org/checkerframework/checker/androidresource/AndroidResourceVisitor.java:
```diff
@@ -0,0 +1,104 @@
+//        System.out.println(node.toString() + ", " + node.getKind() + "\n");
+
+        if (checkAnnotatedTypeHasResOrContainerAnnotations(expressionType) || checkAnnotatedTypeHasResOrContainerAnnotations(variableType)) {
+            checker.report(Result.warning("compound.assignment.not.allowed", kind), node);
```
The error message would be more useful if it contained the specific part that is incorrect and an explanation of what is wrong. The error message just says  `Resource annotation` instead of what the concrete problem is.

aagarwal1012 replied:
I have added a more verbose error message. Please check.

aagarwal1012 committed:
rfac: applied suggestions to AndroidResourceAnnotatedTypeFactory

1. Reduced nesting of the code.
2. Moved expression and identifier variable to `assignContainerAnnotations` method.
3. Changed annotation.equals to AnnotationUtils.areSame

aagarwal1012 committed:
rfac: added a new line in .travis.yml

aagarwal1012 committed:
rfac: applied suggestions to AndroidResourceVisitor

1. Add a variable `hasAnnotation` in `checkAnnotatedTypeHasResOrContainerAnnotations` method.
2. Redefined the warning message given during invalid binary or compound operation.
3. Refined docs.

aagarwal1012 committed:
docs: refined docs

aagarwal1012 committed:
bfix: fixed tests

aagarwal1012 committed:
docs: added docs for annotation classes

aagarwal1012 committed:
Merge remote-tracking branch 'origin/master' into develop

aagarwal1012 committed:
docs: added docs for  method

aagarwal1012 committed:
chor: fixed unused imports

wmdietl reviewed (approved):
This looks good now! We could optimize some of it, but it seems good enough for now.

aagarwal1012 committed:
chor: cleaned build.gradle

aagarwal1012 committed:
chor: beautify code in checkAnnotatedTypeHasResOrContainerAnnotations method

aagarwal1012 commented:
@wmdietl, 
I have made the final changes that you suggested to me.
Thanks, merging this PR.
